File: 0149-max-points-on-a-line/0149-max-points-on-a-line.py
from collections import defaultdict

class Solution:
    def maxPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        res = 0
        for i in range(n):
            slope_count = defaultdict(int)
            for j in range(i + 1, n):
                if points[i][0] == points[j][0]:  # avoid division by zero
                    slope = float('inf')
                else:
                    slope = (points[i][1] - points[j][1]) / (points[i][0] - points[j][0])
                
                slope_count[slope] += 1
            
            res = max(res, max(slope_count.values(), default=0))
        
        return res + 1

File: 0149-max-points-on-a-line/0149-max-points-on-a-line.cpp
class Solution {
public:
    unordered_map<float, int> umap;
    int maxPoints(vector<vector<int>>& points) {
        int n = (int)points.size();
        int res = 0;
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                float key = 0.0;
                if (points[i][0] - points[j][0] == 0) { // avoid division by 0
                    key = INT_MAX;
                }
                else {
                    key = float(points[i][1] - points[j][1]) /
                          float(points[i][0] - points[j][0]);
                }

                umap[key]++;
            }
            for (auto &[_, key] : umap) {
                res = max(res, key);
            }
            umap.clear();
        }
        return res + 1;
    }
};